### PR TITLE
HPCC-23188 Initial work on supporting azure blobs

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by richardkchapman** [comment]
Fails to build on osx:

make[2]: *** No rule to make target `system/azure/azure-storage-cpplite/libazure-storage-lite.so', needed by `Debug/libs/libazurefile.dylib'.  Stop.

File: common/remote/hooks/azure/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+         ${HPCC_SOURCE_DIR}/system/include
+         ${HPCC_SOURCE_DIR}/system/jlib
+         ${HPCC_SOURCE_DIR}/system/azure/azure-storage-cpplite/include
+    )
+
+ADD_DEFINITIONS( -D_USRDLL -DAZURE_FILE_EXPORTS )
+
+HPCC_ADD_LIBRARY( azurefile SHARED ${SRCS}  )
+install ( TARGETS azurefile DESTINATION filehooks )
+target_link_options(azurefile  PUBLIC  "LINKER:-z,defs" )
```
Comment (richardkchapman) [inline common/remote/hooks/azure/CMakeLists.txt:46]: This gives an error on osx (-z option not recognized)

File: system/azure/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        azure-storage-cpplite
+        SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/azure-storage-cpplite
+        BINARY_DIR ${CMAKE_CURRENT_BINARY_DIR}/azure-storage-cpplite
+        CMAKE_ARGS -DCMAKE_CXX_FLAGS=${EXTERNAL_AZURE_CXX_FLAGS} -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTS=OFF -DBUILD_SAMPLES=OFF -DBUILD_SHARED_LIBS=ON
+        BUILD_COMMAND ${CMAKE_MAKE_PROGRAM} LDFLAGS=-Wl,-rpath-link,${LIB_PATH}
+        INSTALL_COMMAND "")
+
+    add_library(azure-storage-lite SHARED IMPORTED GLOBAL)
+    set_property(TARGET azure-storage-lite
+        PROPERTY IMPORTED_LOCATION ${CMAKE_CURRENT_BINARY_DIR}/azure-storage-cpplite/libazure-storage-lite.so )
```
Comment (richardkchapman) [inline system/azure/CMakeLists.txt:41]: Need to use ${CMAKE_SHARED_LIBRARY_SUFFIX) rather than .so

**Comment by ghalliday** [comment]
@AttilaVamos do you know why it is failing for centos 7.6?

**Comment by AttilaVamos** [comment]
@ghalliday I am working on with Ming to find out what is the difference and why it isn't build.

**Comment by AttilaVamos** [comment]
@ghalliday it seems there is some dependency issue or missing a makefile:
In (each time) new and empty build environment  I can build (with make -j24 <module-name>):
- azurefile
- azure-storage-cpplite

but I can't build

- azure-storage-lite

```
make -j24 azure-storage-lite
make: *** No rule to make target 'azure-storage-lite'.  Stop.
```

**Comment by ghalliday** [comment]
azure-storage-lite is a cmake command within a submodule.

system/azure/azure-storage-cpplite contains the cmake file.  It is run as a external project.

**Comment by AttilaVamos** [comment]
@ghalliday in effect there is only one difference between 
```
centos 7.7.1908 (Linux 3.10.0-957.1.3.el7.x86_64)
GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Host: ip-10-20-0-116.ca-central-1.compute.internal
(8 cores)
```
and 
```
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Host: ip-10-20-0-118.ca-central-1.compute.internal
(16 cores)
```
environments and it is the number of building threads, which is 12 for the first and 24 for the second.
When I used 12 in the second environment it is build well.

File: fs/dafsclient/rmtfile.cpp [hunk 3]
```diff
@@ -2024,6 +2024,19 @@ extern DAFSCLIENT_API void installFileHooks(const char *hookFileSpec)
     }
 }
 
+void installDefaultFileHooks(IPropertyTree * config)
+{
+#ifndef _CONTAINERIZED
```
Comment (richardkchapman) [inline fs/dafsclient/rmtfile.cpp:2029]: Why ifndef?

File: ecl/hthor/hthorkey.cpp [hunk 4]
```diff
@@ -51,7 +51,7 @@ static IKeyIndex *openKeyFile(IDistributedFilePart & keyFile)
             if (thissize != -1)
             {
                 StringBuffer remotePath;
-                rfn.getRemotePath(remotePath);
+                rfn.getPath(remotePath);
```
Comment (richardkchapman) [inline ecl/hthor/hthorkey.cpp:54]: Curious why these changes?

File: common/remote/hooks/azure/azurefile.cpp [hunk 5]
```diff
@@ -0,0 +1,827 @@
...
+        ensureMetaData();
+        return fileExists;
+    }
+    virtual bool getTime(CDateTime * createTime, CDateTime * modifiedTime, CDateTime * accessedTime);
+    virtual fileBool isDirectory()
+    {
+        ensureMetaData();
+        if (!fileExists)
+            return notFound;
+        return isDir ? foundYes : foundNo;
```
Comment (richardkchapman) [inline common/remote/hooks/azure/azurefile.cpp:174]: Probably will need changing if my notFound change is merged

File: common/remote/hooks/azure/azurefile.cpp [hunk 6]
```diff
@@ -0,0 +1,827 @@
...
+        return 0;
+    if (pos + len > file->fileSize)
+        len = file->fileSize - pos;
+    if (len == 0)
+        return 0;
+
+    size32_t sizeRead = 0;
+    offset_t lastOffset = pos + len;
+
+    // MORE: Do we ever read file IO from more than one thread?  I'm not convinced we do, and the critical blocks waste space and slow it down.
```
Comment (richardkchapman) [inline common/remote/hooks/azure/azurefile.cpp:308]: I would think (at least in Roxie) that fetch and index reads may well end up reading a shared ifileio from multiple threads.

File: common/remote/hooks/azure/azurefile.cpp [hunk 7]
```diff
@@ -0,0 +1,827 @@
...
+    StringAttr accountName;
+    StringAttr accountKey;
+    std::string containerName;
+    std::string blobName;
+    offset_t fileSize = unknownFileSize;
+    bool haveMeta = false;
+    bool isDir = false;
+    bool fileExists = false;
+    time_t blobModifiedTime = 0;
+    SpinLock lock;
```
Comment (richardkchapman) [inline common/remote/hooks/azure/azurefile.cpp:283]: Delete this

File: ecl/hthor/hthorkey.cpp (hunk 4 above)
Comment (ghalliday) [reply ecl/hthor/hthorkey.cpp:54]: I think they are probably correct - getPath() will return RemotePath if it is remote.  I'm not sure why it is calling getRemotePath().  @jakesmith please review all 3 changes like this.

File: common/remote/hooks/azure/azurefile.cpp [hunk 8]
```diff
@@ -0,0 +1,827 @@
...
+        cred = std::make_shared<azure::storage_lite::shared_key_credential>(accountName, key);
+    }
+    catch (const std::exception & e)
+    {
+        IException * error = makeStringExceptionV(1234, "Azure access: %s", e.what());
+        throw error;
+    }
+
+    std::shared_ptr<azure::storage_lite::storage_account> account = std::make_shared<azure::storage_lite::storage_account>(accountName, cred, /* use_https */ true);
+    return std::make_shared<azure::storage_lite::blob_client>(account, 10);
```
Comment (jakesmith) [inline common/remote/hooks/azure/azurefile.cpp:528]: what it the '10' constant?

File: common/remote/hooks/azure/azurefile.cpp [hunk 9]
```diff
@@ -0,0 +1,827 @@
...
+        if (!streq(ret.error().code.c_str(), "ContainerBeingDeleted"))
+            return true;
+        MilliSleep(100);
+    }
+}
+
+
+void AzureFile::createAppendBlob()
+{
+    //Put blob is used to either create a blob or replace an existing one.
```
Comment (jakesmith) [inline common/remote/hooks/azure/azurefile.cpp:603]: is "Put blob" comment correct?

File: common/remote/hooks/azure/azurefile.cpp [hunk 10]
```diff
@@ -0,0 +1,827 @@
...
+    auto client = getClient();
+    std::istringstream input(std::string((const char *)data, len));
+    //implement append_block_from_buffer based on upload_block_from_buffer
+    auto ret = client->append_block_from_stream(containerName, blobName, input).get();
+    checkError(ret);
+}
+
+void AzureFile::createBlockBlob()
+{
+    //Put blob is used to either create a blob or replace an existing one.
```
Comment (jakesmith) [inline common/remote/hooks/azure/azurefile.cpp:621]: is "Put blob" comment correct?

File: common/remote/hooks/azure/azurefile.cpp [hunk 11]
```diff
@@ -0,0 +1,827 @@
...
+}
+
+
+void AzureFile::appendToBlockBlob(size32_t len, const void * data)
+{
+    auto client = getClient();
+    std::string blockid;
+    auto ret = client->upload_block_from_buffer(containerName, blobName, blockid, (const char *)data, len).get();
+    checkError(ret);
+    //blocks.append(blockid);
```
Comment (jakesmith) [inline common/remote/hooks/azure/azurefile.cpp:638]: worth removing?

File: common/remote/hooks/azure/azurefile.cpp [hunk 12]
```diff
@@ -0,0 +1,827 @@
...
+        fileExists = true;
+        fileSize = properties.size;
+        blobModifiedTime = properties.last_modified;    // Could be more accurate
+        //MORE: extract information from properties.metadata
+    }
+    else
+    {
+        fileExists = false;
+    }
+
```
Comment (jakesmith) [inline common/remote/hooks/azure/azurefile.cpp:737]: extra newline

File: common/remote/hooks/azure/azurefile.cpp [hunk 13]
```diff
@@ -0,0 +1,827 @@
...
+    return new AzureFile(azureFileName);
+}
+
+
+//---------------------------------------------------------------------------------------------------------------------
+
+class AzureFileHook : public CInterfaceOf<IContainedFileHook>
+{
+public:
+    virtual IFile * createIFile(const char *fileName)
```
Comment (jakesmith) [inline common/remote/hooks/azure/azurefile.cpp:763]: override

File: common/remote/hooks/azure/azurefile.cpp [hunk 14]
```diff
@@ -0,0 +1,827 @@
...
+    {
+        if (isAzureFileName(fileName))
+            return createAzureFile(fileName);
+        else
+            return NULL;
```
Comment (jakesmith) [inline common/remote/hooks/azure/azurefile.cpp:768]: nullptr

File: common/remote/hooks/azure/azurefile.cpp [hunk 15]
```diff
@@ -0,0 +1,827 @@
...
+}
+
+MODULE_EXIT()
+{
+    if (azureFileHook)
+    {
+        removeContainedFileHook(azureFileHook);
+        azureFileHook = NULL;
+    }
+    ::Release(azureFileHook);
```
Comment (jakesmith) [inline common/remote/hooks/azure/azurefile.cpp:824]: trivial: move into condition above.

File: common/remote/hooks/libarchive/archive.cpp [hunk 16]
```diff
@@ -39,10 +39,11 @@
  * Installs hooks into createIFile, spotting filenames of the form /my/directory/myfile.zip/{password}/path/within/archive
  */
 
+//Require a trailing / on zip files - use /. for the root - otherwise zip files always get expanded when they are read.
 #ifdef _WIN32
-#define ARCHIVE_SIGNATURE "[.]{zip|tar|tar[.]gz|tgz}{$|/|\\\\}"
+#define ARCHIVE_SIGNATURE "[.]{zip|tar|tar[.]gz|tgz}{/|\\\\}"
 #else
-#define ARCHIVE_SIGNATURE "[.]{zip|tar|tar[.]gz|tgz}{$|/}"
+#define ARCHIVE_SIGNATURE "[.]{zip|tar|tar[.]gz|tgz}/"
```
Comment (jakesmith) [inline common/remote/hooks/libarchive/archive.cpp:46]: JIC - as unrelated(?) is it worth putting this in separate PR?

File: ecl/hthor/hthorkey.cpp (hunk 4 above)
Comment (jakesmith) [reply ecl/hthor/hthorkey.cpp:54]: yes, I don't know why they were calling getRemotePath before, although harmless, it unnecessarily created a full remote path even when local.
All 4 changes look correct afaics.

**Comment by jakesmith** [comment]
@ghalliday looks good, only small minor comments added.

File: common/remote/hooks/libarchive/archive.cpp (hunk 16 above)
Comment (richardkchapman) [reply common/remote/hooks/libarchive/archive.cpp:46]: It's not unrelated... EclWatch fails dealing with zip files (e.g. for ZAP reports) if hooks are installed without this change.

File: cmake_modules/commonSetup.cmake [hunk 17]
```diff
@@ -67,6 +67,7 @@ IF ("${COMMONSETUP_DONE}" STREQUAL "")
   option(USE_OPENSSL "Configure use of OpenSSL" ON)
   option(USE_ZLIB "Configure use of zlib" ON)
   option(USE_CBLAS "Configure use of cblas" ON)
+  option(USE_AZURE "Configure use of azure" ON)
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply cmake_modules/commonSetup.cmake:70]: I don't think so - it's still valid to read Azure blob from bare-metal - either within data center reading remote, or from "lift/shift" cloud installations.

File: common/remote/hooks/azure/azurefile.cpp (hunk 15 above)
Comment (richardkchapman) [reply common/remote/hooks/azure/azurefile.cpp:824]: Where it currently is, azureFileHook will always be null when you call release.

File: common/remote/hooks/azure/azurefile.cpp [hunk 18]
```diff
@@ -0,0 +1,827 @@
...
+constexpr offset_t azureReadRequestSize = 0x400000;  // Default to requesting 4Mb each time
+#endif
+
+MODULE_INIT(INIT_PRIORITY_STANDARD)
+{
+    return true;
+}
+MODULE_EXIT()
+{
+}
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply common/remote/hooks/azure/azurefile.cpp:62]: There seem to be other MODULE blocks further down...

File: common/remote/hooks/azure/azurefile.cpp (hunk 7 above)
Comment (ghalliday) [reply common/remote/hooks/azure/azurefile.cpp]: gone

File: common/remote/hooks/azure/azurefile.cpp [hunk 19]
```diff
@@ -0,0 +1,827 @@
...
+        {
+            size_t copySize = ((lastOffset > endResultOffset) ? endResultOffset : lastOffset) - pos;
+            size_t extractedSize = extractDataFromResult((pos - startResultOffset), copySize, data);
+            assertex(copySize == extractedSize);
+            pos += copySize;
+            len -= copySize;
+            data = (byte *)data + copySize;
+            sizeRead += copySize;
+            if (len == 0)
+                return sizeRead;
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply common/remote/hooks/azure/azurefile.cpp:324]: Worth considering.  I expect this code to be rewritten before the final version.

File: common/remote/hooks/azure/azurefile.cpp [hunk 20]
```diff
@@ -0,0 +1,827 @@
...
+
+offset_t AzureFileAppendBlobWriteIO::appendFile(IFile *file, offset_t pos, offset_t len)
+{
+    UNIMPLEMENTED_X("AzureFileAppendBlobWriteIO::appendFile");
+    return 0;
+}
+
+offset_t AzureFileAppendBlobWriteIO::size()
+{
+    DBGLOG("Warning: Size (%" I64F "u) requested on output IO", offset);
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply common/remote/hooks/azure/azurefile.cpp:441]: It does support it, but it is a bit of a strange thing to request - which is why it is currently logged.  The tracing should probably be removed eventually.  Maybe I should introduce a verbose #define

File: common/remote/hooks/azure/azurefile.cpp (hunk 8 above)
Comment (ghalliday) [reply common/remote/hooks/azure/azurefile.cpp:528]: Very good question - I can't remember.  I'll document it.

File: common/remote/hooks/azure/azurefile.cpp (hunk 15 above)
Comment (ghalliday) [reply common/remote/hooks/azure/azurefile.cpp:824]: copied from archive code.  I'll fix.

File: common/remote/hooks/azure/azurefile.cpp (hunk 6 above)
Comment (ghalliday) [reply common/remote/hooks/azure/azurefile.cpp:308]: The code currently protects for multi threaded access.  If we swap to a streaming interface then the cs shouldn't be necessary.  Can revisit when that works is done.

File: common/remote/hooks/azure/azurefile.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (ghalliday) [reply common/remote/hooks/azure/azurefile.cpp:62]: done

File: common/remote/hooks/azure/azurefile.cpp [hunk 21]
```diff
@@ -0,0 +1,827 @@
...
+class AzureFile;
+class AzureFileReadIO : implements CInterfaceOf<IFileIO>
+{
+public:
+    AzureFileReadIO(AzureFile * _file, const FileIOStats & _stats);
+
+    virtual size32_t read(offset_t pos, size32_t len, void * data) override;
+    virtual offset_t size() override;
+    virtual void close() override
+    {
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply common/remote/hooks/azure/azurefile.cpp:76]: I don't see a big advantage.  I suspect close is never called on files being read anyway.

File: common/remote/hooks/azure/azurefile.cpp (hunk 14 above)
Comment (ghalliday) [reply common/remote/hooks/azure/azurefile.cpp]: done

**Comment by ghalliday** [comment]
@richardkchapman I think I've addressed all the comments.
